discovery/xds/xds.go

```diff
@@ -0,0 +1,187 @@
+type ProtocolVersion string
+
+const (
+	ProtocolV3 = ProtocolVersion("v3")
+)
+
+type HTTPConfig struct {
+	config.HTTPClientConfig `yaml:",inline"`
+	RefreshInterval         model.Duration `yaml:"refresh_interval,omitempty"`
+	FetchTimeout            model.Duration `yaml:"fetch_timeout,omitempty"`
+}
+
+// SDConfig is a base config for xDS-based SD mechanisms
+type SDConfig struct {
+	HTTP   HTTPConfig `yaml:",inline"`
+	Server string     `yaml:"server,omitempty"`
+	// ClientID is sent to the xDS management API to identify this client
+	ClientID string `yaml:"client_id,omitempty"`
+}
+
+func init() {
+	// Register top-level SD Configs
+	discovery.RegisterConfig(&KumaSDConfig{})
+
+	// Register metrics
+	prometheus.MustRegister(kumaFetchDuration)
+	prometheus.MustRegister(kumaFetchSkipUpdateCount)
+	prometheus.MustRegister(kumaFetchFailuresCount)
+
+	// Register protobuf types that need to be marshalled/ unmarshalled
+	// core xDS
+	_ = protoTypes.RegisterMessage((&v3.DiscoveryRequest{}).ProtoReflect().Type())
+	_ = protoTypes.RegisterMessage((&v3.DiscoveryResponse{}).ProtoReflect().Type())
+	// implementations
+	_ = protoTypes.RegisterMessage((&MonitoringAssignment{}).ProtoReflect().Type())
+}
+
+var protoTypes = new(protoregistry.Types)
+
+func protoUnmarshalOptions() proto.UnmarshalOptions {
+	return proto.UnmarshalOptions{
+		DiscardUnknown: true,       // only want known fields
+		Merge:          true,       // always using new messages
+		Resolver:       protoTypes, // only want known types
+	}
+}
+
+func protoJSONUnmarshalOptions() protojson.UnmarshalOptions {
+	return protojson.UnmarshalOptions{
+		DiscardUnknown: true,       // only want known fields
+		Resolver:       protoTypes, // only want known types
+	}
+}
+
+func protoJSONMarshalOptions() protojson.MarshalOptions {
+	return protojson.MarshalOptions{
+		UseProtoNames: true,
+		Resolver:      protoTypes, // only want known types
+	}
+}
+
+type resourceParser func(resources []*anypb.Any, typeUrl string) ([]*targetgroup.Group, error)
+
+// fetchDiscovery implements long-polling via xDS Fetch REST-JSON
+type fetchDiscovery struct {
+	client ResourceClient
+	source string
+
+	refreshInterval time.Duration
+
+	parseResources resourceParser
+	logger         log.Logger
+
+	// metrics
+	// fetchDuration records the duration of a single fetch request, seconds
+	fetchDuration prometheus.Observer
+	// fetchSkipUpdateCount records the number of fetch requests that result in no update
+	fetchSkipUpdateCount prometheus.Counter
+	// fetchFailuresCount records the number of fetch requests that result in an error
+	fetchFailuresCount prometheus.Counter
+}
+
+func (d *fetchDiscovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	defer d.client.Close()
+
+	ticker := time.NewTicker(d.refreshInterval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			level.Debug(d.logger).Log("msg", "polling fetch endpoint", "source", d.source, "client", d.client.ID())
```

Comment:
It was helpful to have a message each time a fetch happened for debugging, but the client ID has no value unless manually set (since it defaults to the hostname) and source currently is always `kuma`. 

I'll remove the client and source for now 👍🏼